schema/encoding/jsonschema.go

```diff
@@ -0,0 +1,169 @@
+	"strconv"
+	"strings"
+)
+
+func boundariesToJSONSchema(bound *schema.Boundaries) (_ []byte, err error) {
+	if bound == nil {
+		return empty, nil
+	}
+	b := new(bytes.Buffer)
+	if _, err := fmt.Fprintf(b, "%q: %s, %q: %s",
+		"minimum", strconv.FormatFloat(bound.Min, 'E', -1, 64),
+		"maximum", strconv.FormatFloat(bound.Max, 'E', -1, 64)); err != nil {
+		return empty, err
+	}
+	return b.Bytes(), nil
+}
+
+// ValidatorToJSONSchema takes a validator and renders to JSON
+func validatorToJSONSchema(v schema.FieldValidator) (_ []byte, err error) {
+	if v == nil {
+		return empty, nil
+	}
+	b := new(bytes.Buffer)
+	switch t := v.(type) {
+	case *schema.String:
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "type", "string"); err != nil {
+			return empty, err
+		}
+		if t.Regexp != "" {
+			if _, err = fmt.Fprintf(b, "%q: %q, ", "pattern", t.Regexp); err != nil {
+				return empty, err
+			}
+		}
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, fmt.Sprintf("%q", value))
+			}
+			if _, err = fmt.Fprintf(b, "%q: [ %s ],", "enum", strings.Join(allowed, ",")); err != nil {
+				return empty, err
+			}
+		}
+		if t.MinLen > 0 {
+			if _, err = fmt.Fprintf(b, "%q: %s, ", "minLength", strconv.FormatInt(int64(t.MinLen), 10)); err != nil {
+				return empty, err
+			}
+		}
+		if t.MaxLen > 0 {
+			if _, err = fmt.Fprintf(b, "%q: %s, ", "maxLength", strconv.FormatInt(int64(t.MaxLen), 10)); err != nil {
+				return empty, err
+			}
+		}
+	case *schema.Integer:
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "type", "number"); err != nil {
+			return empty, err
+		}
+
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, strconv.FormatInt(int64(value), 10))
+			}
+			if _, err = fmt.Fprintf(b, "%q: [ %s ],", "enum", strings.Join(allowed, ",")); err != nil {
+				return empty, err
+			}
+		}
+		bound, err := boundariesToJSONSchema(t.Boundaries)
+		if err != nil {
+			return empty, err
+		}
+		if _, err = b.Write(bound); err != nil {
+			return empty, err
+		}
+		if _, err = b.WriteString(","); err != nil {
+			return empty, err
+		}
+	case *schema.Array:
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "type", "array"); err != nil {
+			return empty, err
+		}
+		if t.ValuesValidator != nil {
+			v, err := validatorToJSONSchema(t.ValuesValidator)
+			if err != nil {
+				return empty, err
+			}
+			if _, err = fmt.Fprintf(b, "%q: %s, ", "items", v); err != nil {
+				return empty, err
+			}
+		}
+	case *schema.Object:
+		buf, _ := SchemaToJSONSchema(t.Schema)
+		if _, err = fmt.Fprintf(b, "%s, ", buf); err != nil {
+			return empty, err
+		}
+	case *schema.Time:
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "type", "string"); err != nil {
+			return empty, err
+		}
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "format", "date-time"); err != nil {
+			return empty, err
+		}
+	default:
+	}
+	b = bytes.NewBuffer(bytes.TrimRight(b.Bytes(), ", "))
```

Comment:
Why are all cases in the switch adding a coma at the end if it is to be trimmed at the end?